File: core/src/main/java/ch/dissem/bitmessage/entity/ObjectMessage.java
/*
 * Copyright 2015 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity;

import ch.dissem.bitmessage.entity.payload.ObjectPayload;
import ch.dissem.bitmessage.entity.payload.ObjectType;
import ch.dissem.bitmessage.entity.payload.Pubkey;
import ch.dissem.bitmessage.entity.valueobject.InventoryVector;
import ch.dissem.bitmessage.entity.valueobject.PrivateKey;
import ch.dissem.bitmessage.exception.ApplicationException;
import ch.dissem.bitmessage.exception.DecryptionFailedException;
import ch.dissem.bitmessage.utils.Bytes;
import ch.dissem.bitmessage.utils.Encode;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import static ch.dissem.bitmessage.utils.Singleton.security;

/**
 * The 'object' command sends an object that is shared throughout the network.
 */
public class ObjectMessage implements MessagePayload {
    private static final long serialVersionUID = 2495752480120659139L;

    private byte[] nonce;
    private long expiresTime;
    private long objectType;
    /**
     * The object's version
     */
    private long version;
    private long stream;

    private ObjectPayload payload;
    private byte[] payloadBytes;

    private ObjectMessage(Builder builder) {
        nonce = builder.nonce;
        expiresTime = builder.expiresTime;
        objectType = builder.objectType;
        version = builder.payload.getVersion();
        stream = builder.streamNumber;
        payload = builder.payload;
    }

    @Override
    public Command getCommand() {
        return Command.OBJECT;
    }

    public byte[] getNonce() {
        return nonce;
    }

    public void setNonce(byte[] nonce) {
        this.nonce = nonce;
    }

    public long getExpiresTime() {
        return expiresTime;
    }

    public long getType() {
        return objectType;
    }

    public ObjectPayload getPayload() {
        return payload;
    }

    public long getVersion() {
        return version;
    }

    public long getStream() {
        return stream;
    }

    public InventoryVector getInventoryVector() {
        return new InventoryVector(
                Bytes.truncate(security().doubleSha512(nonce, getPayloadBytesWithoutNonce()), 32)
        );
    }

    private boolean isEncrypted() {
        return payload instanceof Encrypted && !((Encrypted) payload).isDecrypted();
    }

    public boolean isSigned() {
        return payload.isSigned();
    }

    private byte[] getBytesToSign() {
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            writeHeaderWithoutNonce(out);
            payload.writeBytesToSign(out);
            return out.toByteArray();
        } catch (IOException e) {
            throw new ApplicationException(e);
        }
    }

    public void sign(PrivateKey key) {
        if (payload.isSigned()) {
            payload.setSignature(security().getSignature(getBytesToSign(), key));
        }
    }

    public void decrypt(PrivateKey key) throws IOException, DecryptionFailedException {
        if (payload instanceof Encrypted) {
            ((Encrypted) payload).decrypt(key.getPrivateEncryptionKey());
        }
    }

    public void decrypt(byte[] privateEncryptionKey) throws IOException, DecryptionFailedException {
        if (payload instanceof Encrypted) {
            ((Encrypted) payload).decrypt(privateEncryptionKey);
        }
    }

    public void encrypt(byte[] publicEncryptionKey) throws IOException {
        if (payload instanceof Encrypted) {
            ((Encrypted) payload).encrypt(publicEncryptionKey);
        }
    }

    public void encrypt(Pubkey publicKey) {
        try {
            if (payload instanceof Encrypted) {
                ((Encrypted) payload).encrypt(publicKey.getEncryptionKey());
            }
        } catch (IOException e) {
            throw new ApplicationException(e);
        }
    }

    public boolean isSignatureValid(Pubkey pubkey) throws IOException {
        if (isEncrypted()) throw new IllegalStateException("Payload must be decrypted first");
        return security().isSignatureValid(getBytesToSign(), payload.getSignature(), pubkey);
    }

    @Override
    public void write(OutputStream out) throws IOException {
        if (nonce == null) {
            out.write(new byte[8]);
        } else {
            out.write(nonce);
        }
        out.write(getPayloadBytesWithoutNonce());
    }

    private void writeHeaderWithoutNonce(OutputStream out) throws IOException {
        Encode.int64(expiresTime, out);
        Encode.int32(objectType, out);
        Encode.varInt(version, out);
        Encode.varInt(stream, out);
    }

    public byte[] getPayloadBytesWithoutNonce() {
        try {
            if (payloadBytes == null) {
                ByteArrayOutputStream out = new ByteArrayOutputStream();
                writeHeaderWithoutNonce(out);
                payload.write(out);
                payloadBytes = out.toByteArray();
            }
            return payloadBytes;
        } catch (IOException e) {
            throw new ApplicationException(e);
        }
    }

    public static final class Builder {
        private byte[] nonce;
        private long expiresTime;
        private long objectType = -1;
        private long streamNumber;
        private ObjectPayload payload;

        public Builder nonce(byte[] nonce) {
            this.nonce = nonce;
            return this;
        }

        public Builder expiresTime(long expiresTime) {
            this.expiresTime = expiresTime;
            return this;
        }

        public Builder objectType(long objectType) {
            this.objectType = objectType;
            return this;
        }

        public Builder objectType(ObjectType objectType) {
            this.objectType = objectType.getNumber();
            return this;
        }

        public Builder stream(long streamNumber) {
            this.streamNumber = streamNumber;
            return this;
        }

        public Builder payload(ObjectPayload payload) {
            this.payload = payload;
            if (this.objectType == -1)
                this.objectType = payload.getType().getNumber();
            return this;
        }

        public ObjectMessage build() {
            return new ObjectMessage(this);
        }
    }
}

File: core/src/main/kotlin/ch/dissem/bitmessage/entity/ObjectMessage.kt
/*
 * Copyright 2017 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity

import ch.dissem.bitmessage.entity.payload.ObjectPayload
import ch.dissem.bitmessage.entity.payload.ObjectType
import ch.dissem.bitmessage.entity.payload.Pubkey
import ch.dissem.bitmessage.entity.valueobject.InventoryVector
import ch.dissem.bitmessage.entity.valueobject.PrivateKey
import ch.dissem.bitmessage.exception.ApplicationException
import ch.dissem.bitmessage.exception.DecryptionFailedException
import ch.dissem.bitmessage.utils.Bytes
import ch.dissem.bitmessage.utils.Encode
import ch.dissem.bitmessage.utils.Singleton.cryptography
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream
import java.nio.ByteBuffer
import java.util.*

/**
 * The 'object' command sends an object that is shared throughout the network.
 */
data class ObjectMessage(
    var nonce: ByteArray? = null,
    val expiresTime: Long,
    val payload: ObjectPayload,
    val type: Long,
    /**
     * The object's version
     */
    val version: Long,
    val stream: Long
) : MessagePayload {

    override val command: MessagePayload.Command = MessagePayload.Command.OBJECT

    constructor(
        nonce: ByteArray? = null,
        expiresTime: Long,
        payload: ObjectPayload,
        stream: Long
    ) : this(
        nonce,
        expiresTime,
        payload,
        payload.type?.number ?: throw IllegalArgumentException("payload must have type defined"),
        payload.version,
        stream
    )

    val inventoryVector: InventoryVector
        get() {
            return InventoryVector(Bytes.truncate(cryptography().doubleSha512(
                nonce ?: throw IllegalStateException("nonce must be set"),
                payloadBytesWithoutNonce
            ), 32))
        }

    private val isEncrypted: Boolean
        get() = payload is Encrypted && !payload.isDecrypted

    val isSigned: Boolean
        get() = payload.isSigned

    private val bytesToSign: ByteArray
        get() {
            try {
                val out = ByteArrayOutputStream()
                writer.writeHeaderWithoutNonce(out)
                payload.writer().writeBytesToSign(out)
                return out.toByteArray()
            } catch (e: IOException) {
                throw ApplicationException(e)
            }
        }

    fun sign(key: PrivateKey) {
        if (payload.isSigned) {
            payload.signature = cryptography().getSignature(bytesToSign, key)
        }
    }

    @Throws(DecryptionFailedException::class)
    fun decrypt(key: PrivateKey) {
        if (payload is Encrypted) {
            payload.decrypt(key.privateEncryptionKey)
        }
    }

    @Throws(DecryptionFailedException::class)
    fun decrypt(privateEncryptionKey: ByteArray) {
        if (payload is Encrypted) {
            payload.decrypt(privateEncryptionKey)
        }
    }

    fun encrypt(publicEncryptionKey: ByteArray) {
        if (payload is Encrypted) {
            payload.encrypt(publicEncryptionKey)
        }
    }

    fun encrypt(publicKey: Pubkey) {
        try {
            if (payload is Encrypted) {
                payload.encrypt(publicKey.encryptionKey)
            }
        } catch (e: IOException) {
            throw ApplicationException(e)
        }

    }

    fun isSignatureValid(pubkey: Pubkey): Boolean {
        if (isEncrypted) throw IllegalStateException("Payload must be decrypted first")
        return cryptography().isSignatureValid(bytesToSign, payload.signature ?: return false, pubkey)
    }

    val payloadBytesWithoutNonce: ByteArray by lazy {
        val out = ByteArrayOutputStream()
        writer.writeHeaderWithoutNonce(out)
        payload.writer().write(out)
        out.toByteArray()
    }

    private val writer = Writer(this)
    override fun writer(): StreamableWriter = writer

    private class Writer(
        private val item: ObjectMessage
    ) : StreamableWriter {

        override fun write(out: OutputStream) {
            out.write(item.nonce ?: ByteArray(8))
            out.write(item.payloadBytesWithoutNonce)
        }

        override fun write(buffer: ByteBuffer) {
            buffer.put(item.nonce ?: ByteArray(8))
            buffer.put(item.payloadBytesWithoutNonce)
        }

        internal fun writeHeaderWithoutNonce(out: OutputStream) {
            Encode.int64(item.expiresTime, out)
            Encode.int32(item.type, out)
            Encode.varInt(item.version, out)
            Encode.varInt(item.stream, out)
        }

    }

    class Builder {
        private var nonce: ByteArray? = null
        private var expiresTime: Long = 0
        private var objectType: Long? = null
        private var streamNumber: Long = 0
        private var payload: ObjectPayload? = null

        fun nonce(nonce: ByteArray): Builder {
            this.nonce = nonce
            return this
        }

        fun expiresTime(expiresTime: Long): Builder {
            this.expiresTime = expiresTime
            return this
        }

        fun objectType(objectType: Long): Builder {
            this.objectType = objectType
            return this
        }

        fun objectType(objectType: ObjectType): Builder {
            this.objectType = objectType.number
            return this
        }

        fun stream(streamNumber: Long): Builder {
            this.streamNumber = streamNumber
            return this
        }

        fun payload(payload: ObjectPayload): Builder {
            this.payload = payload
            if (this.objectType == null)
                this.objectType = payload.type?.number
            return this
        }

        fun build(): ObjectMessage {
            return ObjectMessage(
                nonce = nonce,
                expiresTime = expiresTime,
                type = objectType!!,
                version = payload!!.version,
                stream = if (streamNumber > 0) streamNumber else payload!!.stream,
                payload = payload!!
            )
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ObjectMessage) return false
        return expiresTime == other.expiresTime &&
            type == other.type &&
            version == other.version &&
            stream == other.stream &&
            payload == other.payload
    }

    override fun hashCode(): Int {
        var result = Arrays.hashCode(nonce)
        result = 31 * result + (expiresTime xor expiresTime.ushr(32)).toInt()
        result = 31 * result + (type xor type.ushr(32)).toInt()
        result = 31 * result + (version xor version.ushr(32)).toInt()
        result = 31 * result + (stream xor stream.ushr(32)).toInt()
        result = 31 * result + (payload.hashCode())
        return result
    }
}
